[simulation/controller] Implement simulation controller

Dav1dK opened the pull request.

Dav1dK committed:
[simulation/controller] Implement the PhysicsSimulator

Dav1dK committed:
[simulation/controller] Implement MainClass

Dav1dK committed:
[simulation/controller] Implement some parts of SimulationController

Therefore introduce a new Method to SimulationViewInterface, to notify
the controller if the OverflowDialog gets closed.

Dav1dK committed:
[simulation/controller] Make the update() method more efficient

Dav1dK committed:
[simulation/controller] Add background loop for the physics simulation

Dav1dK committed:
[simulation/controller] Update settings if changed by the GUI

Dav1dK committed:
[simulation/controller] Put GUI initialization into a helper method

Dav1dK committed:
[simulation] Remove MenuButtonHandlers, use EventHandlers instead

This saves many classes and interfaces and is thus much simpler.

Dav1dK committed:
[simulation/controller] Initialize the GUI

Dav1dK committed:
[core/model/behavior] Add getTank() to TankAlarm

Dav1dK committed:
[simulation/view] Add show() to SimulationControlInterface

Dav1dK committed:
[simulation/view] Fix calculation of tank fill level

Dav1dK committed:
[simulation/controller] Remove useless package specification

Instead, import the packages.

Dav1dK committed:
[simulation/controller] Create productionSite before the constructor

Dav1dK committed:
[simulation/controller] Fix the default port

It cannot be bigger than 64000.

Dav1dK committed:
[simulation/controller] Make timer attribute final

Dav1dK committed:
[simulation/controller] Implement the remaining parts of the controller

Dav1dK committed:
[core/model/base] Change default pipe threshold

The first two tanks are set to 50%, the other tanks have closed pipes.
Thus this algorithm works for every tankcount greater or equal two.
Therefore increase the minimum tank count to three.

Dav1dK committed:
[simulation/controller] Switch from HashMap to LinkedList

We don't need a hashmap, since we don't observe the model anymore.

Dav1dK committed:
[simulation/controller] Fix checkstyle issues

Dav1dK committed:
[simulation/controller] Cancel tick thread in stop()

Otherwise it will throw a NullPointerException on shutdown.

Dav1dK committed:
[simulation] Fix checkstyle issues

ByteHamster commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/PhysicsSimulator.java:
```diff
@@ -1,42 +1,67 @@
-    public final void setInputTemperature (TankSelector tank, float temperature) {
-        throw new RuntimeException("Not implemented!");
+    public void setInputTemperature(TankSelector tank, float temperature) {
+        productionSite.setInputTemperature(tank, temperature);
```
Ich glaub die Funktion wird nirgends benutzt

ByteHamster replied:
Bitte löschen, wenn das wirklich so ist

Dav1dK replied:
Ich hab die Funktion jetzt benutzt, siehe SimulationController.java:237.

ByteHamster commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+    private final List<TankContainer> tanks = new LinkedList<>();
+
+    private ServerSettingsWrapper settingsWrapper;
+    private static final int DEFAULT_PORT = 12686;
```
Die Konstante bitte irgendwo zentral speichern und an den anderen Stellen löschen

https://github.com/ByteHamster/PSE/blob/master/src/osip-monitoring-view/src/main/java/edu/kit/pse/osip/monitoring/view/settings/GeneralSettings.java#L109

https://github.com/ByteHamster/PSE/blob/master/src/osip-simulation-view/src/main/java/edu/kit/pse/osip/simulation/view/settings/SimulationSettingsWindow.java#L119

ByteHamster commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
-        throw new RuntimeException("Not implemented!");
+    public SimulationController() {
+        try {
+            settingsWrapper = new ServerSettingsWrapper(new File("file"));
```
Da müsste Max noch die Settings anpassen und die Exception entfernen

ByteHamster commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+        controlInterface.setTemperatureListener((tankSelector, number) ->
+                productionSite.setInputTemperature(tankSelector, number.byteValue()));
+        controlInterface.setMotorListener(rpm -> productionSite.getMixTank().getMotor().setRPM(rpm));
+        settingsInterface.setSettingsChangedListener(this::reSetupServer);
```
Die Settings speichern es momentan noch nicht automatisch in den Wrapper ab, soweit ich weiß.

ByteHamster commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
-    public final void update (Observable observable, Object object) {
-        throw new RuntimeException("Not implemented!");
+
+    private class MixTankContainer {
```
Auch ein kurzer Kommentar wäre gut

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
-        throw new RuntimeException("Not implemented!");
+    public SimulationController() {
+        try {
+            settingsWrapper = new ServerSettingsWrapper(new File("file"));
```
Was soll "file" für eine Datei sein? Ich hätte hier eine Standarddatei eingefügt, damit das Programm starten kann!

Dav1dK replied:
Das war nur zum testen. Aber welche Standarddatei? Soll ja auf Linux wie auf Windows Sinn ergeben.

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+
+    private void setupAlarms() {
+        for (TankContainer cont: tanks) {
+            cont.overflowAlarm = new FillAlarm(cont.tank, 95f, AlarmBehavior.GREATER_THAN);
```
Konstanten!

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+    private void setupAlarms() {
+        for (TankContainer cont: tanks) {
+            cont.overflowAlarm = new FillAlarm(cont.tank, 95f, AlarmBehavior.GREATER_THAN);
+            cont.underflowAlarm = new FillAlarm(cont.tank, 5f, AlarmBehavior.SMALLER_THAN);
```
Konstanten!

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+        for (TankContainer cont: tanks) {
+            cont.overflowAlarm = new FillAlarm(cont.tank, 95f, AlarmBehavior.GREATER_THAN);
+            cont.underflowAlarm = new FillAlarm(cont.tank, 5f, AlarmBehavior.SMALLER_THAN);
+            cont.overheatAlarm = new TemperatureAlarm(cont.tank, 5000f, AlarmBehavior.GREATER_THAN);
```
Konstanten!

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+            cont.overflowAlarm = new FillAlarm(cont.tank, 95f, AlarmBehavior.GREATER_THAN);
+            cont.underflowAlarm = new FillAlarm(cont.tank, 5f, AlarmBehavior.SMALLER_THAN);
+            cont.overheatAlarm = new TemperatureAlarm(cont.tank, 5000f, AlarmBehavior.GREATER_THAN);
+            cont.undercoolAlarm = new TemperatureAlarm(cont.tank, 200f, AlarmBehavior.SMALLER_THAN);
```
Konstanten!

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+            cont.undercoolAlarm = new TemperatureAlarm(cont.tank, 200f, AlarmBehavior.SMALLER_THAN);
+        }
+
+        mixCont.overflowAlarm = new FillAlarm(mixCont.tank, 95f, AlarmBehavior.GREATER_THAN);
```
Konstanten!

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+        }
+
+        mixCont.overflowAlarm = new FillAlarm(mixCont.tank, 95f, AlarmBehavior.GREATER_THAN);
+        mixCont.underflowAlarm = new FillAlarm(mixCont.tank, 5f, AlarmBehavior.SMALLER_THAN);
```
Konstanten!

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+
+        mixCont.overflowAlarm = new FillAlarm(mixCont.tank, 95f, AlarmBehavior.GREATER_THAN);
+        mixCont.underflowAlarm = new FillAlarm(mixCont.tank, 5f, AlarmBehavior.SMALLER_THAN);
+        mixCont.overheatAlarm = new TemperatureAlarm(mixCont.tank, 5000f, AlarmBehavior.GREATER_THAN);
```
Konstanten!

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+        mixCont.overflowAlarm = new FillAlarm(mixCont.tank, 95f, AlarmBehavior.GREATER_THAN);
+        mixCont.underflowAlarm = new FillAlarm(mixCont.tank, 5f, AlarmBehavior.SMALLER_THAN);
+        mixCont.overheatAlarm = new TemperatureAlarm(mixCont.tank, 5000f, AlarmBehavior.GREATER_THAN);
+        mixCont.undercoolAlarm = new TemperatureAlarm(mixCont.tank, 200f, AlarmBehavior.SMALLER_THAN);
```
Konstanten!

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
      */
-    private final void checkAlarms () {
-        throw new RuntimeException("Not implemented!");
+    public void startMainLoop() {
```
Warum ist diese Methode public?

Dav1dK replied:
Weil sie im Entwurf public ist. Aber ich kann sie auch private machen.

HansMartinA replied:
Soweit ich weiß, wird sie nicht von außen aufgerufen. Damit gibt es keinen Grund, dass sie public ist.

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
-    public PhysicsSimulator simulator;
-    public TankServer inputServer;
-    public MixTankServer mixServer;
+public class SimulationController extends Application {
```
Vielleicht geht's nur mir so, aber ich finde die Reihenfolge der Methoden unintuitiv und hätte sie anders sortiert (SimulationController(), setupServer(), reSetupServer(), setupAlarms(), startMainLoop(), updateServerValues(), start(), setupView(), stop()).

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+        for (TankContainer cont : tanks) {
+            TankServer old = cont.server;
+            try {
+                cont.server = new TankServer(settingsWrapper.getServerPort(cont.selector, defaultPort++));
```
Müsste nicht anschließend noch ein old.stop() kommen, da ansonsten die Server einfach weiterlaufen?

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,244 @@
+        }
+        MixTankServer old = mixCont.server;
+        try {
+            mixCont.server = new MixTankServer(defaultPort);
```
Hier auch ein old.stop()?

Dav1dK committed:
[simulation/controller] Use PhysicsSimulator to set input temperature

Dav1dK committed:
Store the default port in a central place

Dav1dK committed:
[simulation/controller] Adapt to new SettingsWrapper API

Dav1dK committed:
[simulation/controller] Set changed values in the OPC UA Server

This should happen immediately, not after the next tick cycle.

Dav1dK committed:
[simulation/controller] Improve comment to MixTankContainer

Dav1dK committed:
[simulation/controller] Change method order

Dav1dK committed:
[simulation/controller] Use constants for alarm thresholds

Dav1dK committed:
[simulation/controller] Stop old OPC UA server after setting new ports

Dav1dK committed:
[simulation/controller] Make SimulationController.startMainLoop private

It used to be public because it was public in the design, but is can
easily be private.

Dav1dK committed:
[simulation/controller] Only simulate if there is no overflow

Dav1dK committed:
[core/model/behavior] Pass values between 0 and 1 to TankAlarm

Dav1dK committed:
[simulation/controller] Start OPC UA Servers

Dav1dK committed:
[core/model/base] Set color of mix tank to the mixed color

Dav1dK committed:
[simulation/controller] Chage default config file name

Dav1dK committed:
[simulation/controller] Save changed settings into conf file

Dav1dK committed:
[core/model/behavior] Adapt tests to argument change in dc81030640

Dav1dK committed:
[simulation/controller] Fix checkstyle issues

Dav1dK committed:
[simulation/controller] Stop tick timer prior to deleting the servers

Otherwise it is possible that the next tick has a NPE because the
servers are deleted already.

Dav1dK committed:
[simulation/controller] Close settings dialog after saving the values

Dav1dK commented:
Alle Kritikpuntke sollten gefixt sein. Passt euch der Name des Configfiles?

ByteHamster commented:
Pfad finde ich okay

ByteHamster commented on src/osip-core/src/main/java/edu/kit/pse/osip/core/OSIPConstants.java:
```diff
@@ -0,0 +1,18 @@
+package edu.kit.pse.osip.core;
+
+/**
+ * Some constants which are used in multiple places but doesn't bbelong into the model
```
`doesn't bbelong` ==> `don't belong` (sg/pl vermischt)

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/ProductionSite.java:
```diff
@@ -83,20 +83,19 @@ private void initTanks() {
-                specialPipeAssigned = true;
+            if (count == 0 || count == 1) {
+                threshold = (byte) 50;
+                count++;
```
Count++ würde ich in jedem Durchlauf zählen. Sollte auch so funktionieren, aber ein count, der dann irgendwann aufhört, ist seltsam

Dav1dK replied:
Von mir aus, ich dachte das spart ein kleines bisschen Performance, aber das ist wirklich vernachlässigbar

ByteHamster replied:
Du mit deinen Performance Sachen ^^ Also bevor du sowas machst, dann lieber die OPC UA Sachen nur bei Änderungen setzen :P Das würde mehr helfen

Dav1dK replied:
Also gut, dann mach ich es halt in jedem Durchlauf^^

ByteHamster reviewed (commented):
Sieht recht gut aus vom ersten Anschauen. Ich würde noch auf das control window warten, bevor wir das mergen

HansMartinA commented on src/osip-simulation-controller/src/main/java/edu/kit/pse/osip/simulation/controller/SimulationController.java:
```diff
@@ -1,68 +1,289 @@
-    public SimulationController () {
-        throw new RuntimeException("Not implemented!");
+    public SimulationController() {
+        new File(System.getProperty("user.home") + "/.osip").mkdir();
```
Zur Plattformunabhängigkeit eignet sich hier statt "/" File.separator zu nehmen.

ByteHamster replied:
Wenn du es übrigens so machst, steht da nicht zweimal der gleiche Pfad

```
File settingsLocation = new File(System.getProperty("user.home") + File.separator + ".osip");
settingsLocation.mkdirs();
currentSettings = new ClientSettingsWrapper(new File(settingsLocation, "monitoring.conf"));
```

Dav1dK replied:
Wird gemacht

HansMartinA commented on src/osip-model/src/test/java/edu/kit/pse/osip/core/model/behavior/FillAlarmTest.java:
```diff
@@ -41,7 +41,7 @@ public void init() {
         testLiquid = new Liquid(20f, 300f, new Color(0.5, 0.5, 0.5));
         tank = new Tank(200f, TankSelector.valuesWithoutMix()[0], testLiquid, pipe1, pipe2);
-        alarm = new FillAlarm(tank, 50f, AlarmBehavior.GREATER_THAN);
+        alarm = new FillAlarm(tank, 0.5f, AlarmBehavior.GREATER_THAN);
```
Ich finde, hier ist es besser, auch Konstanten zu benutzen. Damit hättest du nicht x-mal die Werte ändern müssen, sondern nur einmal.

Dav1dK replied:
Finde ich nicht nötig und auch nicht so gut, weil die Liquids in jedem Test neu (und unterschiedlich) erzeugt werden. Denn wenn man die Alarmschwelle ändert, muss man auch oft die Liquids anpassen.

Dav1dK committed:
[simulation/view] Remove SettingsChangedListener

It's not needed since it can be replaced by a simple
EventHandler<ActionEvent>.

Dav1dK committed:
[simulation] Integrate the overflow dialog into the view and controller

Dav1dK committed:
[core] Fix spelling misstake in javadoc

Dav1dK committed:
[core/model/base] Increment counter in every iteration

Thats easier to understand.

Dav1dK committed:
[simulation/controller] Use File.seperator and reduce code duplication

Dav1dK committed:
[simulation/view] Fix class cast exceptions

Dav1dK committed:
[core/model/base] Test temperatures with epsilon value

Dav1dK committed:
[simulation/view] Reset slider textfields if value is illegal

The bidirectional binding woulnd't set the textbox values to allowed
values if an illegal value got entered.

Dav1dK committed:
[simulation] Fix showing the overflow dialog

Dav1dK committed:
[core/model/base] Fix counter for default pipes

Dav1dK committed:
[core/model/base] Use different default temperatures

Dav1dK committed:
[simulation/controller] Print message if simulation gets stopped

For debugging purposes.

Dav1dK committed:
[core/model/base] Set default pipe length to 1

A longer pipe is perplexing as you can't see the content of the pipe.

Dav1dK committed:
[simulation] Fix checkstyle issues

Dav1dK committed:
[simulation/controller] Fix simulation of outgoing pipes

HansMartinA commented:
Wenn ich das Ding ausführe, kommen folgende Fehler vor:
- [x] Die Temperaturslider im Control triggern keine Updates für die Temperaturanzeigen in der Simulation.
- [ ] Wenn es zu einem Overflow kommt, wird das Fenster angezeigt. Wenn ich dann im Control irgendetwas umstelle, d. h. Werte ändere, wird immer wieder ein neues Overflow-Fenster angezeigt.
- [x] Bei einem Reset wird zwar das Model und die Ansicht zurückgesetzt, aber das Control nicht. Auch ein Neustart des Controls, d. h. Schließen und erneut Öffnen, ändert nichts.
- [x] Der Start von OPC UA dauert unter Windows extrem lange (bis zur Anzeige des Fensters bei mir etwa 1 min. und es liegt nicht am Computer), weswegen ein Neustart der Server zu einem Einfrieren der GUI führt.
Ich weiß nicht genau, welche Fehler im Controller liegen und welche woanders. Aber das ist mir erstmal noch aufgefallen.

ByteHamster commented:
> Die Temperaturslider im Control triggern keine Updates für die Temperaturanzeigen in der Simulation.

Doch, tun sie. Die Anzeige ist aber kaputt (geht nur ab 150°) und es dauert eine Weile, bis sich die Temperatur anpasst. Anzeige wird in #293 behoben

> Bei einem Reset wird zwar das Model und die Ansicht zurückgesetzt, aber das Control nicht. Auch ein Neustart des Controls, d. h. Schließen und erneut Öffnen, ändert nichts.

Das muss Niko im control window beheben. Issue ist erstellt #294

ByteHamster commented:
> Wenn es zu einem Overflow kommt, wird das Fenster angezeigt. Wenn ich dann im Control irgendetwas umstelle, d. h. Werte ändere, wird immer wieder ein neues Overflow-Fenster angezeigt.

Da muss Niko noch eine disable methode hinzufügen, die die ganzen Slider deaktiviert. Auch in #294

Dav1dK committed:
[simulation/controller] Fix overflow tank specification
